### PR TITLE
Allow multiple tags in build image command

### PR DESCRIPTION
**Pull request by orzeh**
Several tags for image can be defined by subsequent  calling `withTag()` method of `BuildImageCmd`.

@marcuslinke please check my modifications of Netty `WebTarget` class to support several query parameters with the same name as this class is widely used.

---
This change is [Reviewable](https://reviewable.io/reviews/docker-java/docker-java/739)

### CODE REVIEW

File: src/main/java/com/github/dockerjava/core/command/BuildImageCmdImpl.java [hunk 1]
```diff
@@ -181,7 +190,10 @@ public Long getShmsize() {
     @Override
     public BuildImageCmdImpl withTag(String tag) {
         checkNotNull(tag, "Tag is null");
-        this.tag = tag;
+        if (this.tags == null) {
+            this.tags = new HashSet<>(4);
+        }
+        this.tags.add(tag);
```
Comment (KostyaSha) [inline src/main/java/com/github/dockerjava/core/command/BuildImageCmdImpl.java]: you can't be sure what target API is used to know what to put. Add version check to get conditional and default to latest behaviour.

Comment (orzeh) [inline src/main/java/com/github/dockerjava/core/command/BuildImageCmdImpl.java]: @KostyaSha is such version check already implemented somewhere in the code base or I have to write one?

File: src/main/java/com/github/dockerjava/core/command/BuildImageCmdImpl.java [hunk 2]
```diff
@@ -1,30 +1,30 @@
...
- *
  * Build an image from Dockerfile.
- *
  */
 public class BuildImageCmdImpl extends AbstrAsyncDockerCmd<BuildImageCmd, BuildResponseItem> implements BuildImageCmd {
 
     private InputStream tarInputStream;
 
-    private String tag;
+    private Set<String> tags;
```
Comment (orzeh) [inline src/main/java/com/github/dockerjava/core/command/BuildImageCmdImpl.java]: This is private filed - to preserve backward compatibility I've left `getTag()` method.

Comment (KostyaSha) [inline src/main/java/com/github/dockerjava/core/command/BuildImageCmdImpl.java]: Ah, first of all it `Cmd` in cmd we can break compatibility, but better do .Y increment in version then.

File: src/main/java/com/github/dockerjava/core/command/BuildImageCmdImpl.java (hunk 1 above)
Comment (KostyaSha) [inline src/main/java/com/github/dockerjava/core/command/BuildImageCmdImpl.java]: Check the place where Auth headers are added

File: src/main/java/com/github/dockerjava/core/command/BuildImageCmdImpl.java (hunk 2 above)
Comment (orzeh) [inline src/main/java/com/github/dockerjava/core/command/BuildImageCmdImpl.java]: I don't think this is a breaking change - previously written code will still work with all versions of Docker engine (unless you already called `withTag()` multiple times). 

This change add some features so feel free to release it with incremented .Y in version.

**Comment by orzeh** [comment]
@KostyaSha Is there anything else I should do before it can be merged?

File: src/main/java/com/github/dockerjava/api/command/BuildImageCmd.java [hunk 3]
```diff
@@ -34,7 +34,11 @@
 
     /**
      * "t" in API
+     *
+     * @deprecated since docker API version 1.21 there can be multiple tags
+     * specified so use {@link #getTags()}
      */
+    @Deprecated()
```
Comment (KostyaSha) [inline src/main/java/com/github/dockerjava/api/command/BuildImageCmd.java]: Just `@Deprecated`

File: src/main/java/com/github/dockerjava/core/command/BuildImageCmdImpl.java [hunk 4]
```diff
@@ -84,7 +84,11 @@ public BuildImageCmdImpl(BuildImageCmd.Exec exec, InputStream tarInputStream) {
 
     @Override
     public String getTag() {
-        return tag;
+        if (tags == null || tags.isEmpty()) {
+            return null;
+        }
+        // return first tag to be backward compatible
+        return tags.iterator().next();
```
Comment (KostyaSha) [inline src/main/java/com/github/dockerjava/core/command/BuildImageCmdImpl.java]: So maybe keep both fields and pick List<> and if it null or empty return getTag()?

File: src/test/java/com/github/dockerjava/core/command/BuildImageCmdImplTest.java [hunk 5]
```diff
@@ -274,6 +274,28 @@ public void labels() throws Exception {
...
+    @Test
+    public void multipleTags() throws Exception {
+        if (!getVersion(dockerClient).isGreaterOrEqual(RemoteApiVersion.VERSION_1_21)) {
+            throw new SkipException("API version should be >= 1.21");
+        }
+
+        File baseDir = fileFromBuildTestResource("labels");
+
+        String imageId = dockerClient.buildImageCmd(baseDir).withNoCache(true)
+                .withTag("docker-java-test:tag1").withTag("docker-java-test:tag2")
```
Comment (KostyaSha) [inline src/test/java/com/github/dockerjava/core/command/BuildImageCmdImplTest.java]: imho that looks not obvious. If user wants only one tag it should call `.withTag` if multiple then `.withTags(Set<String>)` Maybe keep `getTag()` and add `getTags()` and prefer getTags with fallback to getTag()?

File: src/main/java/com/github/dockerjava/core/command/BuildImageCmdImpl.java (hunk 4 above)
Comment (orzeh) [inline src/main/java/com/github/dockerjava/core/command/BuildImageCmdImpl.java]: @KostyaSha ok, but then we end up with `withTag(String tag)` and new method `withTags(Set<String> tags)` - I don't think this is more elegant solution. But you decide.

File: src/test/java/com/github/dockerjava/core/command/BuildImageCmdImplTest.java (hunk 5 above)
Comment (orzeh) [inline src/test/java/com/github/dockerjava/core/command/BuildImageCmdImplTest.java]: ok, I'll do it this way :smile:

File: src/main/java/com/github/dockerjava/core/command/BuildImageCmdImpl.java (hunk 4 above)
Comment (KostyaSha) [inline src/main/java/com/github/dockerjava/core/command/BuildImageCmdImpl.java]: if i right understand `witTag(String)` is already existing and we can Deprecate it and remove in future and add new Set?

Comment (orzeh) [inline src/main/java/com/github/dockerjava/core/command/BuildImageCmdImpl.java]: yes

File: src/test/java/com/github/dockerjava/netty/exec/BuildImageCmdExecTest.java [hunk 6]
```diff
@@ -279,6 +279,28 @@ public void labels() throws Exception {
...
+                .withTag("docker-java-test:tag1").withTag("docker-java-test:tag2")
+                .exec(new BuildImageResultCallback())
+                .awaitImageId();
+
+        InspectImageResponse inspectImageResponse = dockerClient.inspectImageCmd(imageId).exec();
+        assertThat(inspectImageResponse, not(nullValue()));
+        LOG.info("Image Inspect: {}", inspectImageResponse.toString());
+
+        assertThat(inspectImageResponse.getRepoTags().size(), equalTo(2));
+        assertThat(inspectImageResponse.getRepoTags().contains("docker-java-test:tag1"), equalTo(true));
```
Comment (KostyaSha) [inline src/test/java/com/github/dockerjava/netty/exec/BuildImageCmdExecTest.java]: if `inspectImageResponse.getRepoTags()` it return List or collection then you need search for hamcrest matcher (probably see example at com/github/dockerjava/core/command/CreateContainerCmdImplTest.java:165 )

**Comment by MarkRx** [comment]
Was this supposed to be closed instead of merged?

**Pull request closed without merging**